File: go/internal/formatters/table.go
package formatters

import (
	"encoding/json"
	"fmt"
	"strings"
)

func FormatTable(data map[string]interface{}, minScore float64, engineURL string) {
	results, _ := data["results"].([]interface{})
	overallScore, _ := data["overall_score"].(float64)
	totalRows, _ := data["total_rows"].(float64)
	if totalRows == 0 {
		totalRows = float64(len(results))
	}
	inputTokens, _ := data["input_tokens"].(float64)
	outputTokens, _ := data["output_tokens"].(float64)
	evalID, _ := data["evaluation_id"].(string)

	failed := 0
	for _, r := range results {
		m, _ := r.(map[string]interface{})
		if passed, _ := m["passed"].(bool); !passed {
			failed++
		}
	}

	scorePct := overallScore * 100
	passed := overallScore >= minScore
	icon := "✅"
	if !passed {
		icon = "❌"
	}
	thresholdPct := minScore * 100

	fmt.Println()
	fmt.Printf("  AI Evaluator — Results\n")
	fmt.Printf("  Overall Score:  %.1f%%  %s %s threshold (%.0f%%)\n", scorePct, icon, map[bool]string{true: "above", false: "below"}[passed], thresholdPct)
	fmt.Printf("  Total rows:     %.0f\n", totalRows)
	fmt.Printf("  Failed:         %d\n", failed)
	fmt.Printf("  Tokens:         ↓%.0f · ↑%.0f\n", inputTokens, outputTokens)
	if evalID != "" {
		fmt.Printf("  Dashboard:      %s/evaluations/%s/report\n", engineURL, evalID)
	}
	fmt.Println()

	sep := strings.Repeat("─", 46)
	fmt.Printf("┌────┬%s┬──────────┬──────┐\n", sep)
	fmt.Printf("│  # │ Query%-42s│ Score    │ Pass │\n", "")
	fmt.Printf("├────┼%s┼──────────┼──────┤\n", sep)

	for i, r := range results {
		m, _ := r.(map[string]interface{})
		query, _ := m["query"].(string)
		if len(query) > 46 {
			query = query[:46]
		}
		scores, _ := m["scores"].(map[string]interface{})
		firstScore := 0.0
		for _, v := range scores {
			if s, ok := v.(float64); ok {
				firstScore = s
				break
			}
		}
		scoreStr := fmt.Sprintf("%.0f%%", firstScore*100)
		passIcon := "✅"
		if p, _ := m["passed"].(bool); !p {
			passIcon = "❌"
		}
		fmt.Printf("│ %-3d│ %-46s│ %-9s│ %s   │\n", i+1, query, scoreStr, passIcon)
	}

	fmt.Printf("└────┴%s┴──────────┴──────┘\n", sep)
	fmt.Println()

	if passed {
		fmt.Printf("✅ Score %.1f%% meets threshold %.2f\n", scorePct, minScore)
	} else {
		fmt.Printf("❌ Score %.1f%% below threshold %.2f\n", scorePct, minScore)
	}
	fmt.Println()
}

func FormatJSON(data map[string]interface{}, minScore float64) string {
	results, _ := data["results"].([]interface{})
	overallScore, _ := data["overall_score"].(float64)
	totalRows, _ := data["total_rows"].(float64)
	if totalRows == 0 {
		totalRows = float64(len(results))
	}
	inputTokens, _ := data["input_tokens"].(float64)
	outputTokens, _ := data["output_tokens"].(float64)
	evalID, _ := data["evaluation_id"].(string)

	failed := 0
	for _, r := range results {
		m, _ := r.(map[string]interface{})
		if passed, _ := m["passed"].(bool); !passed {
			failed++
		}
	}

	out := map[string]interface{}{
		"evaluation_id":  evalID,
		"overall_score":  overallScore,
		"passed":         overallScore >= minScore,
		"min_score":      minScore,
		"total_rows":     int(totalRows),
		"failed_queries": failed,
		"input_tokens":   int(inputTokens),
		"output_tokens":  int(outputTokens),
		"results":        results,
	}
	b, _ := json.MarshalIndent(out, "", "  ")
	return string(b)
}

func FormatJUnit(data map[string]interface{}, minScore float64) string {
	results, _ := data["results"].([]interface{})
	failures := 0
	for _, r := range results {
		m, _ := r.(map[string]interface{})
		if passed, _ := m["passed"].(bool); !passed {
			failures++
		}
	}

	var sb strings.Builder
	sb.WriteString("<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n")
	sb.WriteString(fmt.Sprintf("<testsuite name=\"AI Evaluator\" tests=\"%d\" failures=\"%d\" errors=\"0\" time=\"0\">\n", len(results), failures))

	for i, r := range results {
		m, _ := r.(map[string]interface{})
		query, _ := m["query"].(string)
		if len(query) > 80 {
			query = query[:80]
		}
		query = escapeXML(query)

		if passed, _ := m["passed"].(bool); passed {
			sb.WriteString(fmt.Sprintf("  <testcase classname=\"AI Evaluator\" name=\"Query %d: %s\" time=\"0\">\n", i+1, query))
			sb.WriteString("  </testcase>\n")
		} else {
			scores, _ := m["scores"].(map[string]interface{})
			expected, _ := m["expected_output"].(string)
			got, _ := m["agent_response"].(string)

			sb.WriteString(fmt.Sprintf("  <testcase classname=\"AI Evaluator\" name=\"Query %d: %s\" time=\"0\">\n", i+1, query))
			sb.WriteString(fmt.Sprintf("    <failure message=\"Score below threshold %v\">\n", minScore))
			sb.WriteString(fmt.Sprintf("      Query: %s\n", escapeXML(query)))
			sb.WriteString(fmt.Sprintf("      Expected: %s\n", escapeXML(expected)))
			sb.WriteString(fmt.Sprintf("      Got: %s\n", escapeXML(got)))
			sb.WriteString(fmt.Sprintf("      Scores: %v\n", scores))
			sb.WriteString("    </failure>\n")
			sb.WriteString("  </testcase>\n")
		}
	}

	sb.WriteString("</testsuite>\n")
	return sb.String()
}

func escapeXML(s string) string {
	s = strings.ReplaceAll(s, "&", "&amp;")
	s = strings.ReplaceAll(s, "<", "&lt;")
	s = strings.ReplaceAll(s, ">", "&gt;")
	s = strings.ReplaceAll(s, "\"", "&quot;")
	s = strings.ReplaceAll(s, "'", "&apos;")
	return s
}

File: python/src/aievaluator/formatters/table.py
"""Table formatter using Rich."""

from rich.console import Console
from rich.table import Table


def format_table(data: dict, min_score: float, engine_url: str) -> None:
    """Print evaluation results as a Rich table."""
    console = Console()
    results = data.get("results", [])
    overall_score = data.get("overall_score", 0)
    total_rows = data.get("total_rows", len(results))
    failed = sum(1 for r in results if not r.get("passed", True))
    input_tokens = data.get("input_tokens", 0)
    output_tokens = data.get("output_tokens", 0)
    eval_id = data.get("evaluation_id", "")

    score_pct = overall_score * 100
    passed = overall_score >= min_score
    icon = "✅" if passed else "❌"

    console.print()
    console.print(f"  [bold]AI Evaluator — Results[/bold]")
    console.print(f"  Overall Score:  [bold]{score_pct:.1f}%[/bold]  {icon} {'above' if passed else 'below'} threshold ({min_score*100:.0f}%)")
    console.print(f"  Total rows:     {total_rows}")
    console.print(f"  Failed:         {failed}")
    console.print(f"  Tokens:         ↓{input_tokens:,} · ↑{output_tokens:,}")
    if eval_id:
        console.print(f"  Dashboard:      [link={engine_url}/evaluations/{eval_id}/report]{engine_url}/evaluations/{eval_id}/report[/link]")
    console.print()

    table = Table(show_header=True, header_style="bold")
    table.add_column("#", style="dim", width=4)
    table.add_column("Query", max_width=50)
    table.add_column("Score", justify="right", width=8)
    table.add_column("Pass", justify="center", width=6)

    for i, r in enumerate(results):
        query = r.get("query", "")[:50]
        scores = r.get("scores", {})
        first_score = list(scores.values())[0] if scores else 0
        score_str = f"{first_score * 100:.0f}%"
        passed_icon = "✅" if r.get("passed", True) else "❌"
        table.add_row(str(i + 1), query, score_str, passed_icon)

    console.print(table)
    console.print()

    if passed:
        console.print(f"[green]✅ Score {score_pct:.1f}% meets threshold {min_score}[/green]")
    else:
        console.print(f"[red]❌ Score {score_pct:.1f}% below threshold {min_score}[/red]")
    console.print()
